### PR TITLE
[WIP][LTX Video2Video] start ltx video2video.

### PR DESCRIPTION
**Pull request by sayakpaul**
# What does this PR do?

A bit unsure about how to deal with `conditioning_mask` and whether to consider adding `noise` to `initial_latents` like how it's done here:
https://github.com/huggingface/diffusers/blob/8eb73c872afbe59abab4580aaa591a9851a42e6d/src/diffusers/pipelines/cogvideo/pipeline_cogvideox_video2video.py#L386

<details>
<summary>Code to test</summary>

```py
import torch
from diffusers.pipelines.ltx.pipeline_ltx_video2video import LTXVideoToVideoPipeline
from diffusers.utils import export_to_video, load_video

pipe = LTXVideoToVideoPipeline.from_pretrained("Lightricks/LTX-Video", torch_dtype=torch.bfloat16)
pipe.enable_model_cpu_offload()

input_video = load_video(
    "https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/diffusers/hiker.mp4"
)
prompt = (
    "An astronaut stands triumphantly at the peak of a towering mountain. Panorama of rugged peaks and "
    "valleys. Very futuristic vibe and animated aesthetic. Highlights of purple and golden colors in "
    "the scene. The sky is looks like an animated/cartoonish dream of galaxies, nebulae, stars, planets, "
    "moons, but the remainder of the scene is mostly realistic."
)

video = pipe(
    video=input_video, prompt=prompt, guidance_scale=6, num_inference_steps=50
).frames[0]
export_to_video(video, "output_vid2vid.mp4", fps=24)
```

</details>

Input video:

https://github.com/user-attachments/assets/7cf47adf-cb9e-4e40-b771-0ab8552f277c

Output video:

https://github.com/user-attachments/assets/86722307-9fc1-4be8-b856-8c0cb43bad21

@a-r-r-o-w I am going to mention you in a couple of places where I am unsure about the implementation detail. LMK.

### CODE REVIEW

File: src/diffusers/pipelines/ltx/pipeline_ltx_video2video.py [hunk 1]
```diff
@@ -0,0 +1,859 @@
...
+    def _denormalize_latents(
+        latents: torch.Tensor, latents_mean: torch.Tensor, latents_std: torch.Tensor, scaling_factor: float = 1.0
+    ) -> torch.Tensor:
+        # Denormalize latents across the channel dimension [B, C, F, H, W]
+        latents_mean = latents_mean.view(1, -1, 1, 1, 1).to(latents.device, latents.dtype)
+        latents_std = latents_std.view(1, -1, 1, 1, 1).to(latents.device, latents.dtype)
+        latents = latents * latents_std / scaling_factor + latents_mean
+        return latents
+
+    def prepare_latents(
```
Comment (sayakpaul) [inline src/diffusers/pipelines/ltx/pipeline_ltx_video2video.py:447]: @a-r-r-o-w this implementation.

File: src/diffusers/pipelines/ltx/pipeline_ltx_video2video.py [hunk 2]
```diff
@@ -0,0 +1,859 @@
...
+            num_inference_steps,
+            device,
+            timesteps,
+            sigmas=sigmas,
+            mu=mu,
+        )
+        num_warmup_steps = max(len(timesteps) - num_inference_steps * self.scheduler.order, 0)
+        latent_sigma = torch.tensor(
+            sigmas[:1].repeat(batch_size * num_videos_per_prompt), dtype=prompt_embeds.dtype, device=device
+        )
```
Comment (sayakpaul) [inline src/diffusers/pipelines/ltx/pipeline_ltx_video2video.py:731]: Is this the right way to pass it? @a-r-r-o-w

File: src/diffusers/pipelines/ltx/pipeline_ltx_video2video.py [hunk 3]
```diff
@@ -0,0 +1,860 @@
...
+                latents = self._unpack_latents(
+                    latents,
+                    latent_num_frames,
+                    latent_height,
+                    latent_width,
+                    self.transformer_spatial_patch_size,
+                    self.transformer_temporal_patch_size,
+                )
+
+                noise_pred = noise_pred[:, :, 1:]
```
Comment (a-r-r-o-w) [inline src/diffusers/pipelines/ltx/pipeline_ltx_video2video.py:811]: This was only need for image-to-video because we don't want to denoise the first frame (because it is the actual encoded image latent itself)

File: src/diffusers/pipelines/ltx/pipeline_ltx_video2video.py [hunk 4]
```diff
@@ -0,0 +1,860 @@
...
+    def attention_kwargs(self):
+        return self._attention_kwargs
+
+    @property
+    def interrupt(self):
+        return self._interrupt
+
+    @torch.no_grad()
+    @replace_example_docstring(EXAMPLE_DOC_STRING)
+    def __call__(
```
Comment (a-r-r-o-w) [inline src/diffusers/pipelines/ltx/pipeline_ltx_video2video.py:541]: I don't see a `strength` parameter here. Without strength, it is not possible to control how much effect of the original video you would like to have in the output video. Low strength -> low denoising steps -> more similar video. And high strength -> more denoising steps -> less similar video. This is more of a naive approach that we use for vid-to-vid, but with techniques like RF-inversion and FlowEdit (which we'll directly add in modular diffusers instead of pipelines), the quality and possibilities are endless!

File: src/diffusers/pipelines/ltx/pipeline_ltx_video2video.py (hunk 2 above)
Comment (a-r-r-o-w) [reply src/diffusers/pipelines/ltx/pipeline_ltx_video2video.py:732]: This looks incorrect to me. I don't think you have to change sigmas here. You will have to calculate the timesteps based on the `strength` parameter. This would be the `latent_timestep` you see in the other vid-to-vid pipelines. Based on the latent timestep, the amount of noise that should be added to original video latent will be determined. The noise addition process happens via the `scale_noise` method in flow matching schedulers I think (the method name is different from the normal diffusion schedulers)

File: src/diffusers/pipelines/ltx/pipeline_ltx_video2video.py [hunk 5]
```diff
@@ -0,0 +1,860 @@
...
+                retrieve_latents(self.vae.encode(vid.unsqueeze(0).permute(0, 2, 1, 3, 4)), generator) for vid in video
+            ]
+
+        init_latents = torch.cat(init_latents, dim=0).to(dtype)
+        init_latents = self._normalize_latents(init_latents, self.vae.latents_mean, self.vae.latents_std)
+        # `ones()` as we want to condition on all?
+        conditioning_mask = torch.ones(mask_shape, device=device, dtype=dtype)
+
+        noise = randn_tensor(shape, generator=generator, device=device, dtype=dtype)
+        # TODO: consider adding the noise w.r.t the flow equation? CogVideoX vid2vid
```
Comment (a-r-r-o-w) [inline src/diffusers/pipelines/ltx/pipeline_ltx_video2video.py:505]: The `scale_noise` method in the scheduler handles this

File: src/diffusers/pipelines/ltx/pipeline_ltx_video2video.py [hunk 6]
```diff
@@ -0,0 +1,860 @@
...
+            ]
+        else: # `premute()` because we want `batch_size, num_channels, num_frames, height, width`
+            init_latents = [
+                retrieve_latents(self.vae.encode(vid.unsqueeze(0).permute(0, 2, 1, 3, 4)), generator) for vid in video
+            ]
+
+        init_latents = torch.cat(init_latents, dim=0).to(dtype)
+        init_latents = self._normalize_latents(init_latents, self.vae.latents_mean, self.vae.latents_std)
+        # `ones()` as we want to condition on all?
+        conditioning_mask = torch.ones(mask_shape, device=device, dtype=dtype)
```
Comment (a-r-r-o-w) [inline src/diffusers/pipelines/ltx/pipeline_ltx_video2video.py:502]: `conditioning_mask` will not be required anywhere in the video-to-video pipeline I think. It was only required in image-to-video because we wanted first latent frame to be actual image latent, and all the other latent frames to be pure noise. Probably did not require it as a simple `torch.cat` would have sufficed, but it is what it is now haha

**Review by a-r-r-o-w — COMMENTED** [review]
Hope this answer some of your questions! Happy to help with anything else :hugs:

**Comment by sayakpaul** [comment]
@a-r-r-o-w thanks for your comments. Didn't know about `scale_noise()` :D 

I have applied the changes you proposed and I am running a sweep over params. Results: https://wandb.ai/sayakpaul/ltx_video2video/runs/pc4lfivy

<details>
<summary>Sweep</summary>

```py
import torch
from diffusers.pipelines.ltx.pipeline_ltx_video2video import LTXVideoToVideoPipeline
from diffusers.utils import export_to_video, load_video
import wandb

pipe = LTXVideoToVideoPipeline.from_pretrained("Lightricks/LTX-Video", torch_dtype=torch.bfloat16).to("cuda")

input_video = load_video(
    "https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/diffusers/hiker.mp4"
)
prompt = (
    "An astronaut stands triumphantly at the peak of a towering mountain. Panorama of rugged peaks and "
    "valleys. Very futuristic vibe and animated aesthetic. Highlights of purple and golden colors in "
    "the scene. The sky is looks like an animated/cartoonish dream of galaxies, nebulae, stars, planets, "
    "moons, but the remainder of the scene is mostly realistic."
)
wandb.init(project="ltx_video2video")

filenames = []
for s in [1.0, 0.8, 0.7, 0.9]:
    for steps in [50, 60, 70]:
        for cfg in [5, 6, 7, 8]:
            video_name = f"strength@{s}_steps@{steps}_cfg@{cfg}.mp4"
            video = pipe(
                video=input_video, prompt=prompt, guidance_scale=cfg, num_inference_steps=steps, strength=s
            ).frames[0]
            export_to_video(video, video_name, fps=24)
            wandb.log(
                {"videos": wandb.Video(video_name, caption=video_name.replace(".mp4", ""), fps=24)}
            )
```

</details>

LMK if you have any concerns on the implementation.

**Comment by sayakpaul** [comment]
@a-r-r-o-w LMK if the current implementation needs further changes when you get a moment.

**Comment by sayakpaul** [comment]
Also ccing @yoavhacohen in case you wanna review.

**Comment by a-r-r-o-w** [comment]
Not super convinced about the quality of the naive vid2vid approach tbh so I'm unsure if we should proceed to add it. Actually, it doesn't really seem to work convincingly for flow-matching based video models from what I tried in Comfy. Newer techniques like FlowEdit or RF-Inversion work better (working on it for modular-diffusers), so maybe we could tackle it there, but the tradeoff is that you need to perform some number of inversion steps whereas here there is no inversion which makes it lightweight.

**Comment by sayakpaul** [comment]
> Actually, it doesn't really seem to work convincingly for flow-matching based video models from what I tried in Comfy. 

This is signal enough for us to not proceed I guess. Closing this but keeping the branch.
